src/black/__init__.py

```diff
@@ -6424,11 +6424,20 @@ def diff(a: str, b: str, a_name: str, b_name: str) -> str:
     """Return a unified diff string between strings `a` and `b`."""
     import difflib
 
-    a_lines = [line + "\n" for line in a.splitlines()]
-    b_lines = [line + "\n" for line in b.splitlines()]
-    return "".join(
-        difflib.unified_diff(a_lines, b_lines, fromfile=a_name, tofile=b_name, n=5)
-    )
+    a_lines = [line for line in a.splitlines(keepends=True)]
+    b_lines = [line for line in b.splitlines(keepends=True)]
+    diff_lines = []
+    for line in difflib.unified_diff(
+        a_lines, b_lines, fromfile=a_name, tofile=b_name, n=5
+    ):
+        # Work around https://bugs.python.org/issue2142
+        # See https://www.gnu.org/software/diffutils/manual/html_node/Incomplete-Lines.html
+        if line[-1] == "\n":
+            diff_lines.append(line)
+        else:
+            diff_lines.append(line + "\n")
+            diff_lines.append("\\ No newline at end of file\n")
+    return "".join(diff_lines)
```

Comment:
Taking a second look at this, this is confusing IMO. `\\ No newline at end of file\n` means to me that the else block should only happen at the end of the diff, but the code is written like this condition can happen wherever in the diff. I would appreciate if this could be clarified. If this extra code is only useful for diff's ending, would be it possible to just check and modify the diff's ending like this (just illustrative code):
```py3
diff_lines = unified_diff(a, b, ...)
last_line = diff_lines[-1]
if last_line[-1] != "\n":
    diff_lines[-1] = last_line + "\n"
    diff_lines.append("\\ No newline at end of file\n")
```

If it's the other way around, could the `\\ No newline at end of file\n` message be tweaked?

Reply:
Per [the spec](https://www.gnu.org/software/diffutils/manual/html_node/Incomplete-Lines.html), only the last line of a file can be the _incomplete line_, yes.

However, if you pass in non-final lines without newlines, `difflib` will accept that, and happily produce output lines without newlines.

Since that latter effect is really Garbage-In-Garbage-Out, and we're using `keepends` now to ensure that cannot happen, your suggested change makes sense, yes.

Reply:
Actually, wait. `diff_lines` doesn't necessarily have the last line of the input file as its last line. In our case, since we always add a final newline, it's only the last `-...` line that may hit this case, not the last line in `diff_lines`, which will be the corresponding `+...` line.

See tests/data/missing_final_newline.diff for this exact case.

In the general case, a _context_ line (` ...`) or added line (`+ ...`) could require this marker too.

Reply:
Technically, I could walk backwards in `diff_lines`, checking the final `-`, final `+`, and final ` ` for missing final newline (and if it's the ` ` case, barfing if we also see it on a `+` or `-` case) but if I was going to do that, I should also be checking for a `\` line, which would appear if BPO 2142 is fixed, and hence this workaround would be deactivated in that case.

Reply:
If the text of the `\...` line is confusing, that can be trivially changed, this is just copied from the spec, but in the actual implementation of `diff`, it's localised with gettext. (Although I fear that there are patch-parsers in the wild which are looking for _this text_ rather than the `\` marker.)

We could even look at the first character of the affected line, and have the text vary by whether it was a `-...`, `+...`, or ` ...` line.

I won't do that without some positive direction though, as it feels like it's adding needless complexity for what is a pretty-standard text string, i.e. `git diff` also produces this literal string. Even GitHub's diff preview uses the exact string "No newline at end of file" in its popup. And it's complexity that a future fix for BPO 2142 probably won't have, since the existing proposed fixes didn't try to distinguish this either. So the patch format would revert back to what it is here, at that point.

Reply:
Alright that makes sense. Thanks for describing the specification!